symtabAPI/src/Object-elf.C

```diff
@@ -1566,6 +1575,68 @@ void ObjectELF::load_object(bool alloc_syms) {
+            // Attributes are encoded in a vendor-specific section of type
+            // SHT_RISCV_ATTRIBUTES and name .riscv.attributes.
+
+            // The attribute data contains a series of <tag, value> pair.
+            // The tag is a uleb128 integer indication the attribute tag. The
+            // value is the attribute value corresponding to the tag. The RISC-V
+            // ABI defines that if the tag is odd, the attribute value will be
+            // an integer. Otherwise, it will be a string. 
+
+            // See https://github.com/bminor/binutils-gdb/blob/master/binutils/readelf.c
+            size_t elf_len;
+            const char *riscv_attr_addr = elfHdr->e_rawfile(elf_len) + riscv_attr_addr_;
+            int riscv_attr_size = riscv_attr_size_;
+            auto riscv_parse_attr_int = [this](int tag, int value) -> bool {
+                switch (tag) {
+                    case RiscvAttrTag::unaligned_access:
+                        riscv_attrs.unaligned_access = boost::make_optional(value != 0);
+                        return false;
+                    case RiscvAttrTag::stack_align:
+                        riscv_attrs.stack_align = boost::make_optional(value);
+                        return false;
+                    case RiscvAttrTag::priv_spec:
+                        riscv_attrs.priv_spec = boost::make_optional(value);
+                        return false;
+                    case RiscvAttrTag::priv_spec_minor:
+                        riscv_attrs.priv_spec_minor = boost::make_optional(value);
+                        return false;
+                    case RiscvAttrTag::priv_spec_revision:
+                        riscv_attrs.priv_spec_revision = boost::make_optional(value);
+                        return false;
+                    case RiscvAttrTag::atomic_abi:
+                        riscv_attrs.atomic_abi = boost::make_optional(value);
+                        return false;
+                    case RiscvAttrTag::x3_reg_usage:
+                        riscv_attrs.stack_align = boost::make_optional(value);
+                        return false;
+                    default:
+                        return true;
+                }
+            };
+            auto riscv_parse_attr_string = [this](int tag, std::string value) -> bool {
+                switch (tag) {
+                    case RiscvAttrTag::arch:
+                        riscv_attrs.riscv_extension_string = value;
+                        return false;
+                    default:
+                        return true;
+                }
+            };
+            bool result = parse_attrs(riscv_attr_addr,
+                    riscv_attr_size,
+                    "riscv",
+                    riscv_parse_attr_int,
+                    riscv_parse_attr_string);
+            if (!result) {
+                create_printf("%s[%d]: riscv attributes missing or corrupted\n", FILE__, __LINE__);
```

Comment:
Will `get_riscv_extensions` work if `parse_attrs` fails? If not, then we should probably return early here.

Reply:
Agree, let's return early.